dra-evolution/testdata/classes.yaml

```diff
@@ -1,162 +1,208 @@
+kind: ResourceClass
 metadata:
   name: example.com-barzer
-  labels:
-    vendor: example.com
-    device-type: gpu
-spec:
-  driver: example.com-barzer
+request:
+  requirements:
+  - device:
+      driverName: example.com-barzer
 ---
 # Any example.com GPU, even though Foozer and
-# Barzer GPUs use different drivers.
-apiVersion: devmgmtproto.k8s.io/v1alpha1
-kind: DeviceClass
+# Barzer GPUs use different drivers. In this case,
+# "example.com GPU" is defined as "a GPU provided
+# by these two drivers" (no standardization and
+# no need for the driver vendors to cooperate).
+apiVersion: resource.k8s.io/v1alpha2
+kind: ResourceClass
 metadata:
   name: example.com-gpu
-spec:
-  selector:
-    matchLabels:
-      vendor: example.com
-      device-type: gpu
+request:
+  requirements:
+  - device:
+      selector: "device.driverName == 'example.com-foozer' || device.driverName == 'example.com-barzer'"
 ---
 # An example.com SR-IOV NIC
-apiVersion: devmgmtproto.k8s.io/v1alpha1
-kind: DeviceClass
+apiVersion: resource.k8s.io/v1alpha2
+kind: ResourceClass
 metadata:
   name: example.com-sriov-nic
-  labels:
-    vendor: example.com
-    device-type: sriov-nic
-spec:
-  driver: example.com-sriov
+request:
+  requirements:
+  - device:
+      driverName: example.com-sriov
 ---
 # An example.org SR-IOV NIC (different vendor)
-apiVersion: devmgmtproto.k8s.io/v1alpha1
-kind: DeviceClass
+apiVersion: resource.k8s.io/v1alpha2
+kind: ResourceClass
 metadata:
   name: example.org-sriov-nic
-  labels:
-    vendor: example.org
-    device-type: sriov-nic
-spec:
-  driver: example.org-sriov
+request:
+  requirements:
+  - device:
+      driverName: example.org-sriov
 ---
 # Request any SR-IOV NIC. This will allow the claim to be satisfied by
-# example.com or example.org NICs.
-apiVersion: devmgmtproto.k8s.io/v1alpha1
-kind: DeviceClass
+# example.com or example.org NICs. In contrast to the example.com-gpu above,
+# here it is assumed that "SR-IOV NIC" is a concept defined in a KEP which
+# then also standardizes an attribute that vendors can use to declare
+# that their device satisfies the requirements in that KEP. All of the
+# NIC examples use attribute names from that fictional KEP.
+apiVersion: resource.k8s.io/v1alpha2
+kind: ResourceClass
 metadata:
   name: sriov-nic
-spec:
-  selector:
-    matchLabels:
-      device-type: sriov-nic
+request:
+  requirements:
+  - device:
+      selector: "'type.nic.k8s.io' in device.attributes" # Doesn't care whether it's "pf" or "vf".
 ---
 # Example.com SR-IOV NIC on a 1Gbps link
-apiVersion: devmgmtproto.k8s.io/v1alpha1
-kind: DeviceClass
+apiVersion: resource.k8s.io/v1alpha2
+kind: ResourceClass
 metadata:
   name: example.com-sriov-nic-1g
-spec:
-  driver: example.com-sriov
-  constraints: "device.bandwidth == '1G'"
+request:
+  requirements:
+  - device:
+      driverName: example.com-sriov
+      selector: "device.quantityAttributes['bandwidth.nic.k8s.io'].compareTo(quantity('1G')) == 0"
```

Comment:
Why do we need `device.quantityAttributes['bandwidth.nic.k8s.io'].compareTo(quantity('1G')) == 0`? Is that really how K8s CEL works? Why not at least just `device.quantityAttributes['bandwidth.nic.k8s.io'] == quantity('1G')`. It's still hideous due to the compile time type-safety requirement (which I still want to understand better), but it's a *little* better.

Reply:
> Is that really how K8s CEL works?

Yes. I asked. For reference: https://kubernetes.io/docs/reference/using-api/cel/#kubernetes-quantity-library

> compile time type-safety requirement (which I still want to understand better)

It's not a hard requirement, just something that I found useful because a validation error when creating an object is easier to detect and deal with than a runtime error that occurs each time the scheduler evaluates the expression for a node.

We now have `device.attributes["bandwidth.nic.k8s.io"]` for those who don't care about that safety net.

Reply:
To be continued in https://github.com/kubernetes-sigs/wg-device-management/pull/14#discussion_r1614903726